File: axiom_backend/ai_researcher/core_rag/chunker.py
import re
import os
from typing import List, Dict, Any, Optional

class Chunker:
    """
    Splits Markdown text into overlapping chunks based on section boundaries and token budget.

    Features:
    - Structure-aware chunking respecting section boundaries
    - Token budget instead of fixed paragraph count
    - Hierarchical title extraction for embedding context
    - Recursive semantic splitting for oversized chunks
    """
    def __init__(
        self,
        max_chunk_tokens: int = 512,
        overlap_tokens: int = 64,
        min_chunk_tokens: int = 50,
        # Legacy support
        paragraphs_per_chunk: int = 2,
        overlap_paragraphs: int = 1
    ):
        # Determine chunking mode
        self._use_token_chunking = os.getenv("CHUNKER_MODE", "token").lower() != "paragraph"

        if self._use_token_chunking:
            self.max_chunk_tokens = max_chunk_tokens
            self.overlap_tokens = overlap_tokens
            self.min_chunk_tokens = min_chunk_tokens
            logger_msg = f"Token-based chunking: max={max_chunk_tokens}, overlap={overlap_tokens}, min={min_chunk_tokens}"
        else:
            # Legacy paragraph-based mode
            if overlap_paragraphs >= paragraphs_per_chunk:
                raise ValueError("Overlap paragraphs must be less than paragraphs per chunk.")
            self.paragraphs_per_chunk = paragraphs_per_chunk
            self.overlap_paragraphs = overlap_paragraphs
            logger_msg = f"Paragraph-based chunking: {paragraphs_per_chunk} per chunk, {overlap_paragraphs} overlap"

        # Use regex to split by one or more newlines, keeping separators
        self._paragraph_split_pattern = re.compile(r'(\n\s*\n+)')
        # Pattern to detect markdown headings
        self._heading_pattern = re.compile(r'^(#{1,6})\s+(.+)$', re.MULTILINE)
        # Pattern to detect markdown images
        self._image_pattern = re.compile(r'!\[([^\]]*)\]\(([^\)]+)\)')

        print(f"Chunker initialized: {logger_msg}")

    def _count_tokens(self, text: str) -> int:
        """Approximate token count using word count * 1.3 ratio."""
        # Simple approximation: words * 1.3 for English text
        # This is faster than tiktoken and good enough for chunking
        return int(len(text.split()) * 1.3)

    def _extract_images_from_text(self, text: str) -> List[Dict[str, Any]]:
        """Extract image references from markdown text."""
        images = []
        for match in self._image_pattern.finditer(text):
            images.append({
                "alt_text": match.group(1),
                "path": match.group(2),
                "position": match.start()
            })
        return images

    def _extract_heading_hierarchy(self, markdown_content: str) -> Dict[int, List[str]]:
        """
        Extract heading hierarchy for each paragraph position.

        Returns:
            Dict mapping paragraph_index -> List[ancestor headings]
        """
        # Split into paragraphs
        parts = self._paragraph_split_pattern.split(markdown_content)
        paragraphs = []
        current_paragraph = ""

        for part in parts:
            if part:
                if self._paragraph_split_pattern.match(part):
                    if current_paragraph:
                        current_paragraph += part
                        paragraphs.append(current_paragraph.strip())
                    current_paragraph = ""
                else:
                    current_paragraph += part
        if current_paragraph.strip():
            paragraphs.append(current_paragraph.strip())

        # Track heading hierarchy
        heading_stack = []  # List of (level, title)
        paragraph_to_headings = {}

        for i, para in enumerate(paragraphs):
            lines = para.strip().split('\n')
            first_line = lines[0].strip() if lines else ""

            # Check if this paragraph starts with a heading
            heading_match = self._heading_pattern.match(first_line)
            if heading_match:
                level = len(heading_match.group(1))
                title = heading_match.group(2).strip()

                # Pop lower-level headings from stack
                while heading_stack and heading_stack[-1][0] >= level:
                    heading_stack.pop()

                # Add this heading to stack
                heading_stack.append((level, title))

            # Store current hierarchy for this paragraph
            paragraph_to_headings[i] = [h[1] for h in heading_stack]

        return paragraph_to_headings

    def _recursive_split(self, text: str, max_tokens: int, overlap_tokens: int) -> List[str]:
        """
        Recursively split oversized text using semantic boundaries.

        Separators hierarchy:
        1. \n\n (paragraph break)
        2. \n (line break)
        3. . (sentence end)
        4. ; or , (clause break)
        5. Binary split at midpoint
        """
        if self._count_tokens(text) <= max_tokens:
            return [text]

        separators = [
            ("\n\n", 2),   # Paragraph break
            ("\n", 1),      # Line break
            (". ", 2),      # Sentence end
            ("; ", 2),      # Clause break
            (", ", 2),      # Clause break
        ]

        for sep, sep_tokens in separators:
            if sep in text:
                parts = text.split(sep)
                if len(parts) > 1:
                    # Find split point closest to midpoint within budget
                    mid_idx = len(parts) // 2

                    # Try to find a good split point
                    left_text = ""
                    right_text = ""

                    for i in range(mid_idx, 0, -1):
                        candidate = sep.join(parts[:i])
                        if self._count_tokens(candidate) <= max_tokens:
                            left_text = candidate
                            right_text = sep.join(parts[i:])
                            break

                    if not left_text:
                        # No good split found, try from beginning
                        for i in range(1, len(parts)):
                            candidate = sep.join(parts[:i])
                            if self._count_tokens(candidate) > max_tokens:
                                left_text = sep.join(parts[:i-1]) if i > 1 else parts[0]
                                right_text = sep.join(parts[max(1, i-1):])
                                break

                    if left_text and right_text:
                        # Add overlap to right chunk
                        if overlap_tokens > 0:
                            left_words = left_text.split()
                            overlap_words = left_words[-min(len(left_words), overlap_tokens):]
                            right_text = " ".join(overlap_words) + " " + right_text

                        # Recursively split if still too large
                        result = []
                        for chunk in self._recursive_split(left_text, max_tokens, overlap_tokens):
                            result.append(chunk)
                        for chunk in self._recursive_split(right_text, max_tokens, overlap_tokens):
                            result.append(chunk)
                        return result

        # Fallback: binary split at word midpoint
        words = text.split()
        if len(words) > max_tokens:
            mid = len(words) // 2
            left = " ".join(words[:mid])
            right = " ".join(words[mid - overlap_tokens:] if overlap_tokens > 0 else words[mid:])

            result = []
            for chunk in self._recursive_split(left, max_tokens, overlap_tokens):
                result.append(chunk)
            for chunk in self._recursive_split(right, max_tokens, overlap_tokens):
                result.append(chunk)
            return result

        return [text]

    def chunk(self, markdown_content: str, doc_metadata: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """
        Chunks the provided Markdown content using structure-aware token-based chunking.

        Args:
            markdown_content: The full Markdown text of the document.
            doc_metadata: Optional dictionary containing document-level metadata
                          (e.g., doc_id) to be added to each chunk's metadata.

        Returns:
            A list of chunk dictionaries, where each dictionary contains
            the chunk 'text' and its 'metadata'.
        """
        if not markdown_content:
            return []

        # Extract heading hierarchy for all paragraphs
        paragraph_to_headings = self._extract_heading_hierarchy(markdown_content)

        # Split into paragraphs
        parts = self._paragraph_split_pattern.split(markdown_content)
        paragraphs = []
        current_paragraph = ""

        for part in parts:
            if part:
                if self._paragraph_split_pattern.match(part):
                    if current_paragraph:
                        current_paragraph += part
                        paragraphs.append(current_paragraph.strip())
                    current_paragraph = ""
                else:
                    current_paragraph += part
        if current_paragraph.strip():
            paragraphs.append(current_paragraph.strip())

        if not paragraphs:
            return []

        # Choose chunking mode
        if self._use_token_chunking:
            return self._chunk_token_based(paragraphs, paragraph_to_headings, doc_metadata)
        else:
            return self._chunk_paragraph_based(paragraphs, doc_metadata)

    def _chunk_token_based(
        self,
        paragraphs: List[str],
        paragraph_to_headings: Dict[int, List[str]],
        doc_metadata: Optional[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """Token-based structure-aware chunking."""
        chunks = []
        doc_id = doc_metadata.get("doc_id", "unknown_doc") if doc_metadata else "unknown_doc"
        chunk_id_counter = 0

        current_chunk_paras = []
        current_chunk_tokens = 0
        current_start_idx = 0
        heading_stack = []

        for i, para in enumerate(paragraphs):
            para_tokens = self._count_tokens(para)
            para_headings = paragraph_to_headings.get(i, [])

            # Check if this is a heading paragraph
            first_line = para.strip().split('\n')[0] if para.strip() else ""
            is_heading = self._heading_pattern.match(first_line)

            # Determine if we should start a new chunk
            should_start_new = False

            if is_heading:
                # New section - emit current chunk if not empty
                if current_chunk_paras:
                    should_start_new = True
                # Update heading stack for new section
                heading_stack = para_headings
            elif current_chunk_tokens + para_tokens > self.max_chunk_tokens:
                # Would exceed budget - emit current chunk
                if current_chunk_paras:
                    should_start_new = True

            if should_start_new and current_chunk_paras:
                # Emit current chunk
                chunk_text = "\n\n".join(current_chunk_paras)
                chunk_tokens = self._count_tokens(chunk_text)

                # Check if chunk needs recursive splitting
                if chunk_tokens > self.max_chunk_tokens:
                    sub_chunks = self._recursive_split(
                        chunk_text, self.max_chunk_tokens, self.overlap_tokens
                    )
                    for sub_text in sub_chunks:
                        chunks.append(self._create_chunk(
                            sub_text, doc_id, chunk_id_counter,
                            current_start_idx, current_start_idx + len(current_chunk_paras) - 1,
                            heading_stack, doc_metadata
                        ))
                        chunk_id_counter += 1
                else:
                    chunks.append(self._create_chunk(
                        chunk_text, doc_id, chunk_id_counter,
                        current_start_idx, current_start_idx + len(current_chunk_paras) - 1,
                        heading_stack, doc_metadata
                    ))
                    chunk_id_counter += 1

                # Start new chunk with overlap
                if self.overlap_tokens > 0 and current_chunk_paras:
                    # Take last portion for overlap
                    overlap_text = current_chunk_paras[-1] if current_chunk_paras else ""
                    overlap_words = overlap_text.split()
                    if len(overlap_words) > self.overlap_tokens:
                        overlap_para = " ".join(overlap_words[-self.overlap_tokens:])
                        current_chunk_paras = [overlap_para]
                        current_chunk_tokens = self._count_tokens(overlap_para)
                    else:
                        current_chunk_paras = []
                        current_chunk_tokens = 0
                else:
                    current_chunk_paras = []
                    current_chunk_tokens = 0

                current_start_idx = i

            # Add paragraph to current chunk
            current_chunk_paras.append(para)
            current_chunk_tokens += para_tokens

            # Update heading stack if not a heading itself
            if not is_heading:
                heading_stack = para_headings

        # Emit final chunk
        if current_chunk_paras:
            chunk_text = "\n\n".join(current_chunk_paras)
            chunk_tokens = self._count_tokens(chunk_text)

            # Merge into previous if too small
            if chunk_tokens < self.min_chunk_tokens and chunks:
                # Append to last chunk
                last_chunk = chunks[-1]
                last_chunk["text"] = last_chunk["text"] + "\n\n" + chunk_text
                last_chunk["metadata"]["end_paragraph_index"] = current_start_idx + len(current_chunk_paras) - 1
                last_chunk["metadata"]["token_count"] = self._count_tokens(last_chunk["text"])
            else:
                # Check if needs recursive splitting
                if chunk_tokens > self.max_chunk_tokens:
                    sub_chunks = self._recursive_split(
                        chunk_text, self.max_chunk_tokens, self.overlap_tokens
                    )
                    for sub_text in sub_chunks:
                        chunks.append(self._create_chunk(
                            sub_text, doc_id, chunk_id_counter,
                            current_start_idx, current_start_idx + len(current_chunk_paras) - 1,
                            heading_stack, doc_metadata
                        ))
                        chunk_id_counter += 1
                else:
                    chunks.append(self._create_chunk(
                        chunk_text, doc_id, chunk_id_counter,
                        current_start_idx, current_start_idx + len(current_chunk_paras) - 1,
                        heading_stack, doc_metadata
                    ))

        return chunks

    def _chunk_paragraph_based(
        self,
        paragraphs: List[str],
        doc_metadata: Optional[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """Legacy paragraph-based chunking for backward compatibility."""
        chunks = []
        doc_id = doc_metadata.get("doc_id", "unknown_doc") if doc_metadata else "unknown_doc"
        chunk_id_counter = 0

        step = self.paragraphs_per_chunk - self.overlap_paragraphs
        for i in range(0, len(paragraphs), step):
            end_index = i + self.paragraphs_per_chunk
            current_paragraphs = paragraphs[i:end_index]

            if not current_paragraphs:
                continue

            chunk_text = "\n\n".join(current_paragraphs)
            chunks.append(self._create_chunk(
                chunk_text, doc_id, chunk_id_counter,
                i, min(end_index, len(paragraphs)) - 1,
                [], doc_metadata  # No section titles in legacy mode
            ))
            chunk_id_counter += 1

            if end_index >= len(paragraphs):
                break

        return chunks

    def _create_chunk(
        self,
        text: str,
        doc_id: str,
        chunk_index: int,
        start_para: int,
        end_para: int,
        section_titles: List[str],
        doc_metadata: Optional[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Create a chunk dictionary with all metadata."""
        image_refs = self._extract_images_from_text(text)

        chunk_meta = {
            "doc_id": doc_id,
            "chunk_id": f"{doc_id}_chunk_{chunk_index:04d}",
            "chunk_index": chunk_index,
            "start_paragraph_index": start_para,
            "end_paragraph_index": end_para,
            "section_titles": section_titles,
            "token_count": self._count_tokens(text),
            "image_refs": image_refs
        }

        if doc_metadata:
            chunk_meta.update({k: v for k, v in doc_metadata.items() if k != "doc_id"})

        return {
            "text": text,
            "metadata": chunk_meta
        }

File: axiom_backend_ng/internal/chunker/chunker.go
// Package chunker is the Go port of axiom_backend/ai_researcher/core_rag/chunker.py.
// It splits a markdown document into overlapping, section-aware,
// token-budgeted chunks matching the Python implementation's behaviour
// closely enough that the resulting chunk_ids can slot into the
// existing documents schema.
//
// Port status: the token-based chunking path (the production default)
// is ported in full. Paragraph-mode is intentionally omitted — it is a
// legacy compatibility path and the migration plan drops it.
package chunker

import (
	"fmt"
	"regexp"
	"strings"
)

// DefaultConfig matches the Python Chunker defaults in
// core_rag/chunker.py:21-29.
func DefaultConfig() Config {
	return Config{
		MaxChunkTokens: 512,
		OverlapTokens:  64,
		MinChunkTokens: 50,
	}
}

// Config tunes the chunker. See core_rag/chunker.py:21-29.
type Config struct {
	MaxChunkTokens int
	OverlapTokens  int
	MinChunkTokens int
}

// Chunk is the Go equivalent of a chunker.py chunk dict. Text is the
// chunk body and Metadata carries the ancillary fields the Python
// version emits: chunk_id, chunk_index, paragraph range, section
// hierarchy, image refs, page range.
type Chunk struct {
	Text     string
	Metadata ChunkMetadata
}

// ChunkMetadata mirrors the keys the Python _create_chunk helper
// populates. Keys not represented here ("doc_id") live on the parent
// Chunk via the DocID passed to Chunk().
type ChunkMetadata struct {
	DocID               string     `json:"doc_id"`
	ChunkID             string     `json:"chunk_id"`
	ChunkIndex          int        `json:"chunk_index"`
	StartParagraphIndex int        `json:"start_paragraph_index"`
	EndParagraphIndex   int        `json:"end_paragraph_index"`
	SectionTitles       []string   `json:"section_titles"`
	TokenCount          int        `json:"token_count"`
	ImageRefs           []ImageRef `json:"image_refs"`
	PageStart           string     `json:"page_start"`
	PageEnd             string     `json:"page_end"`
	// Extras are pass-through fields from doc_metadata (minus the keys
	// Python excludes). Preserved so callers can propagate title,
	// authors, etc. through to the chunk row without changing the
	// metadata schema.
	Extras map[string]any `json:"-"`
}

// ImageRef is one markdown image reference inside a chunk.
type ImageRef struct {
	AltText  string `json:"alt_text"`
	Path     string `json:"path"`
	Position int    `json:"position"`
}

// Regexes — compiled once and reused across calls. Expressions ported
// verbatim from the Python module.
var (
	pageMarkerRe       = regexp.MustCompile(`^\{(\d+)\}-{10,}$`)
	paragraphSplitRe   = regexp.MustCompile(`(\n\s*\n+)`)
	imageRe            = regexp.MustCompile(`!\[([^\]]*)\]\(([^\)]+)\)`)
	headingFirstLineRe = regexp.MustCompile(`^(#{1,6})\s+(.+)$`)
)

// Chunker holds one configured instance. Safe for concurrent use — the
// config fields are read-only after construction.
type Chunker struct {
	cfg Config
}

// New returns a Chunker populated with cfg. Zero-value fields fall back
// to DefaultConfig values.
func New(cfg Config) *Chunker {
	def := DefaultConfig()
	if cfg.MaxChunkTokens <= 0 {
		cfg.MaxChunkTokens = def.MaxChunkTokens
	}
	if cfg.OverlapTokens < 0 {
		cfg.OverlapTokens = def.OverlapTokens
	}
	if cfg.MinChunkTokens <= 0 {
		cfg.MinChunkTokens = def.MinChunkTokens
	}
	return &Chunker{cfg: cfg}
}

// DocMetadata is the subset of document-level fields the chunker
// consumes. DocID is required for chunk_id derivation; PageLabelMap
// translates Marker physical page indices to logical labels; Extras
// are merged into every emitted chunk's metadata.
type DocMetadata struct {
	DocID        string
	PageLabelMap map[int]string
	Extras       map[string]any
}

// Chunk splits markdown into structure-aware chunks. Matches
// chunker.py::Chunker.chunk.
func (c *Chunker) Chunk(markdown string, doc DocMetadata) []Chunk {
	if markdown == "" {
		return nil
	}

	// First pass: full-content heading hierarchy.
	paragraphs := splitParagraphs(markdown)
	if len(paragraphs) == 0 {
		return nil
	}
	headings := extractHeadingHierarchy(paragraphs)

	// Strip Marker page markers and build paragraph → page mapping.
	clean, paraToPage := stripPageMarkers(paragraphs, doc.PageLabelMap)
	if len(clean) == 0 {
		return nil
	}

	// Re-extract headings if the stripping changed paragraph indices.
	if len(clean) != len(paragraphs) {
		headings = extractHeadingHierarchy(clean)
	}

	return c.chunkTokenBased(clean, headings, paraToPage, doc)
}

// chunkTokenBased is the port of Chunker._chunk_token_based.
// Intentionally preserves the Python logic's sequence of decisions so
// the resulting chunks line up closely with recorded golden outputs.
func (c *Chunker) chunkTokenBased(
	paragraphs []string,
	paragraphHeadings map[int][]string,
	paragraphToPage map[int]string,
	doc DocMetadata,
) []Chunk {
	docID := doc.DocID
	if docID == "" {
		docID = "unknown_doc"
	}
	var (
		chunks        []Chunk
		current       []string
		currentTokens int
		startIdx      int
		headingStack  []string
		pageStart     = pageAt(paragraphToPage, 0, "1")
		pageEnd       = pageStart
		chunkCounter  int
	)

	emit := func(endIdx int) {
		if len(current) == 0 {
			return
		}
		text := strings.Join(current, "\n\n")
		tokens := countTokens(text)
		if tokens > c.cfg.MaxChunkTokens {
			for _, sub := range c.recursiveSplit(text) {
				chunks = append(chunks, c.buildChunk(sub, docID, chunkCounter, startIdx, endIdx, headingStack, pageStart, pageEnd, doc))
				chunkCounter++
			}
		} else {
			chunks = append(chunks, c.buildChunk(text, docID, chunkCounter, startIdx, endIdx, headingStack, pageStart, pageEnd, doc))
			chunkCounter++
		}
	}

	for i, para := range paragraphs {
		paraTokens := countTokens(para)
		paraHeadings := paragraphHeadings[i]
		firstLine := firstNonEmptyLine(para)
		isHeading := headingFirstLineRe.MatchString(firstLine)

		shouldFlush := false
		if isHeading && len(current) > 0 {
			shouldFlush = true
			headingStack = paraHeadings
		} else if currentTokens+paraTokens > c.cfg.MaxChunkTokens && len(current) > 0 {
			shouldFlush = true
		}

		if shouldFlush {
			emit(startIdx + len(current) - 1)

			// Reset page tracking.
			pageStart = pageAt(paragraphToPage, i, pageEnd)
			pageEnd = pageStart

			// Start next chunk with optional overlap from the tail of
			// the previous one.
			if c.cfg.OverlapTokens > 0 && len(current) > 0 {
				last := current[len(current)-1]
				words := strings.Fields(last)
				if len(words) > c.cfg.OverlapTokens {
					overlap := strings.Join(words[len(words)-c.cfg.OverlapTokens:], " ")
					current = []string{overlap}
					currentTokens = countTokens(overlap)
				} else {
					current = nil
					currentTokens = 0
				}
			} else {
				current = nil
				currentTokens = 0
			}
			startIdx = i
		}

		current = append(current, para)
		currentTokens += paraTokens

		// Track page range for the current chunk. The Python code uses
		// len(current)==1 to seed pageStart at the first paragraph.
		pPage := pageAt(paragraphToPage, i, pageEnd)
		if len(current) == 1 {
			pageStart = pPage
		}
		pageEnd = pPage

		if !isHeading {
			headingStack = paraHeadings
		}
	}

	// Final chunk — maybe merge into the previous one if undersized.
	if len(current) > 0 {
		text := strings.Join(current, "\n\n")
		tokens := countTokens(text)
		endIdx := startIdx + len(current) - 1

		if tokens < c.cfg.MinChunkTokens && len(chunks) > 0 {
			last := &chunks[len(chunks)-1]
			last.Text = last.Text + "\n\n" + text
			last.Metadata.EndParagraphIndex = endIdx
			last.Metadata.TokenCount = countTokens(last.Text)
			last.Metadata.PageEnd = pageEnd
		} else if tokens > c.cfg.MaxChunkTokens {
			for _, sub := range c.recursiveSplit(text) {
				chunks = append(chunks, c.buildChunk(sub, docID, chunkCounter, startIdx, endIdx, headingStack, pageStart, pageEnd, doc))
				chunkCounter++
			}
		} else {
			chunks = append(chunks, c.buildChunk(text, docID, chunkCounter, startIdx, endIdx, headingStack, pageStart, pageEnd, doc))
		}
	}
	return chunks
}

// buildChunk mirrors Chunker._create_chunk.
func (c *Chunker) buildChunk(
	text, docID string, index int,
	startPara, endPara int,
	sectionTitles []string,
	pageStart, pageEnd string,
	doc DocMetadata,
) Chunk {
	refs := extractImageRefs(text)
	meta := ChunkMetadata{
		DocID:               docID,
		ChunkID:             fmt.Sprintf("%s_chunk_%04d", docID, index),
		ChunkIndex:          index,
		StartParagraphIndex: startPara,
		EndParagraphIndex:   endPara,
		SectionTitles:       append([]string{}, sectionTitles...),
		TokenCount:          countTokens(text),
		ImageRefs:           refs,
		PageStart:           pageStart,
		PageEnd:             pageEnd,
	}
	if len(doc.Extras) > 0 {
		meta.Extras = make(map[string]any, len(doc.Extras))
		for k, v := range doc.Extras {
			if k == "doc_id" || k == "page_label_map" {
				continue
			}
			meta.Extras[k] = v
		}
	}
	return Chunk{Text: text, Metadata: meta}
}

// recursiveSplit is the port of Chunker._recursive_split. Separators
// hierarchy and overlap semantics match the Python implementation.
func (c *Chunker) recursiveSplit(text string) []string {
	if countTokens(text) <= c.cfg.MaxChunkTokens {
		return []string{text}
	}
	type sep struct {
		s string
	}
	separators := []sep{
		{"\n\n"}, {"\n"}, {". "}, {"; "}, {", "},
	}
	for _, sp := range separators {
		if !strings.Contains(text, sp.s) {
			continue
		}
		parts := strings.Split(text, sp.s)
		if len(parts) <= 1 {
			continue
		}
		mid := len(parts) / 2
		var leftText, rightText string
		for i := mid; i > 0; i-- {
			candidate := strings.Join(parts[:i], sp.s)
			if countTokens(candidate) <= c.cfg.MaxChunkTokens {
				leftText = candidate
				rightText = strings.Join(parts[i:], sp.s)
				break
			}
		}
		if leftText == "" {
			for i := 1; i < len(parts); i++ {
				candidate := strings.Join(parts[:i], sp.s)
				if countTokens(candidate) > c.cfg.MaxChunkTokens {
					if i > 1 {
						leftText = strings.Join(parts[:i-1], sp.s)
					} else {
						leftText = parts[0]
					}
					start := i - 1
					if start < 1 {
						start = 1
					}
					rightText = strings.Join(parts[start:], sp.s)
					break
				}
			}
		}
		if leftText != "" && rightText != "" {
			if c.cfg.OverlapTokens > 0 {
				leftWords := strings.Fields(leftText)
				take := c.cfg.OverlapTokens
				if take > len(leftWords) {
					take = len(leftWords)
				}
				overlap := strings.Join(leftWords[len(leftWords)-take:], " ")
				rightText = overlap + " " + rightText
			}
			out := c.recursiveSplit(leftText)
			out = append(out, c.recursiveSplit(rightText)...)
			return out
		}
	}
	// Fallback: binary split at word midpoint.
	words := strings.Fields(text)
	if len(words) > c.cfg.MaxChunkTokens {
		midW := len(words) / 2
		left := strings.Join(words[:midW], " ")
		var right string
		if c.cfg.OverlapTokens > 0 && midW-c.cfg.OverlapTokens >= 0 {
			right = strings.Join(words[midW-c.cfg.OverlapTokens:], " ")
		} else {
			right = strings.Join(words[midW:], " ")
		}
		out := c.recursiveSplit(left)
		out = append(out, c.recursiveSplit(right)...)
		return out
	}
	return []string{text}
}

// countTokens approximates token count with the same len(.split) * 1.3
// heuristic Python uses.
func countTokens(s string) int {
	return int(float64(len(strings.Fields(s))) * 1.3)
}

// splitParagraphs reproduces the Python split logic.
func splitParagraphs(md string) []string {
	// The Python code uses re.split with a capturing group so separators
	// are retained, then joins each paragraph with its trailing
	// whitespace before stripping. We reimplement that here.
	idx := paragraphSplitRe.FindAllStringIndex(md, -1)
	if len(idx) == 0 {
		if s := strings.TrimSpace(md); s != "" {
			return []string{s}
		}
		return nil
	}
	var paragraphs []string
	current := ""
	cursor := 0
	for _, m := range idx {
		start, end := m[0], m[1]
		// text between cursor and start is the paragraph body; the
		// match itself is the separator.
		current += md[cursor:start]
		// The Python version appends the separator onto the paragraph
		// before stripping, so effectively the paragraph body is
		// whatever preceded the separator.
		if stripped := strings.TrimSpace(current); stripped != "" {
			paragraphs = append(paragraphs, stripped)
		}
		current = ""
		cursor = end
	}
	current += md[cursor:]
	if stripped := strings.TrimSpace(current); stripped != "" {
		paragraphs = append(paragraphs, stripped)
	}
	return paragraphs
}

// extractHeadingHierarchy ports _extract_heading_hierarchy.
func extractHeadingHierarchy(paragraphs []string) map[int][]string {
	stack := [][2]string{} // level(as single-char int) + title
	out := make(map[int][]string, len(paragraphs))
	for i, para := range paragraphs {
		first := firstNonEmptyLine(para)
		if m := headingFirstLineRe.FindStringSubmatch(first); m != nil {
			level := len(m[1])
			title := strings.TrimSpace(m[2])
			for len(stack) > 0 && len(stack[len(stack)-1][0]) >= level {
				stack = stack[:len(stack)-1]
			}
			stack = append(stack, [2]string{strings.Repeat("#", level), title})
		}
		titles := make([]string, 0, len(stack))
		for _, s := range stack {
			titles = append(titles, s[1])
		}
		out[i] = titles
	}
	return out
}

// stripPageMarkers removes Marker page-break paragraphs (e.g.
// "{3}------------") and returns the clean list plus paragraph→page
// mapping. labelMap is optional; missing entries default to
// str(physical_idx + 1) for parity with the Python fallback.
func stripPageMarkers(paragraphs []string, labelMap map[int]string) ([]string, map[int]string) {
	clean := make([]string, 0, len(paragraphs))
	paraToPage := make(map[int]string, len(paragraphs))
	currentPage := "1"
	for _, para := range paragraphs {
		if m := pageMarkerRe.FindStringSubmatch(strings.TrimSpace(para)); m != nil {
			var physical int
			if _, err := fmt.Sscanf(m[1], "%d", &physical); err != nil {
				continue
			}
			if v, ok := labelMap[physical]; ok {
				currentPage = v
			} else {
				currentPage = fmt.Sprintf("%d", physical+1)
			}
			continue
		}
		paraToPage[len(clean)] = currentPage
		clean = append(clean, para)
	}
	return clean, paraToPage
}

// extractImageRefs ports _extract_images_from_text.
func extractImageRefs(text string) []ImageRef {
	matches := imageRe.FindAllStringSubmatchIndex(text, -1)
	if len(matches) == 0 {
		return nil
	}
	out := make([]ImageRef, 0, len(matches))
	for _, m := range matches {
		out = append(out, ImageRef{
			AltText:  text[m[2]:m[3]],
			Path:     text[m[4]:m[5]],
			Position: m[0],
		})
	}
	return out
}

func firstNonEmptyLine(s string) string {
	lines := strings.Split(strings.TrimSpace(s), "\n")
	if len(lines) == 0 {
		return ""
	}
	return strings.TrimSpace(lines[0])
}

func pageAt(m map[int]string, i int, def string) string {
	if v, ok := m[i]; ok {
		return v
	}
	return def
}
